modules/videoio/test/test_audio.cpp

```diff
@@ -0,0 +1,253 @@
+        throw SkipTestException(backend.first + " backend was not found");
+
+    doTest();
+}
+
+INSTANTIATE_TEST_CASE_P(/**/, Audio, testing::ValuesIn(audioParams));
+
+class MediaTestFixture : public AudioBaseTest, public testing::TestWithParam <paramCombination>
+{
+public:
+    MediaTestFixture():
+        videoType(get<3>(GetParam())),
+        height(get<4>(GetParam())),
+        width(get<5>(GetParam())),
+        numberOfFrames(get<6>(GetParam())),
+        fps(get<7>(GetParam())),
+        psnrThreshold(get<8>(GetParam()))
+        {
+            format = get<0>(GetParam());
+            expectedNumAudioCh = get<1>(GetParam());
+            epsilon = get<2>(GetParam());
+            backend = get<9>(GetParam());
+            root = "audio/";
+            fileName = "test_audio";
+            params = {  CAP_PROP_AUDIO_STREAM, 0,
+                        CAP_PROP_VIDEO_STREAM, 0,
+                        CAP_PROP_AUDIO_DATA_DEPTH, CV_16S };
+        };
+    void doTest()
+    {
+        getValidAudioData();
+        getValidVideoData();
+        readFile();
+        checkAudio();
+        checkVideoFrames();
+    }
+
+private:
+    void getValidVideoData()
+    {
+        Mat img(height, width, videoType);
+        for (int i = 0; i < numberOfFrames; ++i)
+        {
+            generateFrame(i, numberOfFrames, img);
+            validVideoData.push_back(img);
+        }
+    }
+    void readFile()
+    {
+        ASSERT_TRUE(cap.open(findDataFile(root + fileName + "." + format), backend.second, params));
+
+        const int audioBaseIndex = static_cast<int>(cap.get(cv::CAP_PROP_AUDIO_BASE_INDEX));
+        const int numberOfChannels = (int)cap.get(CAP_PROP_AUDIO_TOTAL_CHANNELS);
+        ASSERT_EQ(expectedNumAudioCh, numberOfChannels);
+
+        const int samplePerSecond = (int)cap.get(CAP_PROP_AUDIO_SAMPLES_PER_SECOND);
+        ASSERT_EQ(44100, samplePerSecond);
+        int samplesPerFrame = (int)(1./fps*samplePerSecond);
+        int audioSamplesTolerance = samplesPerFrame / 2;
+
+        double f = 0;
+        audioData.resize(numberOfChannels);
+        for (;;)
+        {
+            if (cap.grab())
+            {
+                SCOPED_TRACE(cv::format("frame=%d", (int)(videoData.size()+1)));
+                ASSERT_TRUE(cap.retrieve(videoFrame));
+                for (int nCh = 0; nCh < numberOfChannels; nCh++)
+                {
+                    ASSERT_TRUE(cap.retrieve(audioFrame, audioBaseIndex+nCh));
+                    for (int i = 0; i < audioFrame.cols; i++)
+                    {
+                        f = audioFrame.at<signed short>(0,i) / 32768.0;
+                        audioData[nCh].push_back(f);
+                    }
+                }
+                ASSERT_LT(abs(cap.get(CAP_PROP_AUDIO_POS) -  (cap.get(CAP_PROP_POS_MSEC)/ 1000  + 1./fps) * samplePerSecond), audioSamplesTolerance);
+                ASSERT_LT(abs(audioFrame.cols - samplesPerFrame), audioSamplesTolerance);
```

Comment:
`audioFrame.cols` must be same across channels.

`audioFrame.cols` should not be checked for the first frame (or multiple no-audio frames in the beginning) and the last frame (due to the last frame synchronization, which reads audio till the end).